spec/integration/slices_spec.rb

```diff
@@ -81,6 +81,188 @@ class Slice < Hanami::Slice
+        require "hanami"
+
+        module Main
+          class Slice < Hanami::Slice
+            config.actions.format :html
+          end
+        end
+      RUBY
+
+      write "slices/main/config/slice.rb", <<~RUBY
+        module Main
+          class Slice < Hanami::Slice
+            config.actions.format :json
+          end
+        end
+      RUBY
+
+      require "hanami/prepare"
+
+      expect(Hanami.app.slices[:main]).to be Main::Slice
+      expect(Hanami.app.slices[:main].config.actions.format).to eq([:html])
+    end
+  end
+
+  it "Loading a nested slice with its own defined slice class" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/app.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "slices/main/slices/nested/config/slice.rb", <<~RUBY
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :json
+            end
+          end
+        end
+      RUBY
+
+      require "hanami/prepare"
+
+      expect(Hanami.app.slices[:main]).to be Main::Slice
+      expect(Hanami.app.slices[:main].slices[:nested].config.actions.format).to eq([:json])
+    end
+  end
+
+  it "Loading a nested slice with its own defined slice class prefers the parent slice's defined slice class" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/app.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "slices/main/config/slices/nested.rb", <<~RUBY
+        require "hanami"
+
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :html
+            end
+          end
+        end
+      RUBY
+
+      write "slices/main/slices/nested/config/slice.rb", <<~RUBY
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :json
+            end
+          end
+        end
+      RUBY
+
+      require "hanami/prepare"
+
+      expect(Hanami.app.slices[:main]).to be Main::Slice
+      expect(Hanami.app.slices[:main].slices[:nested].config.actions.format).to eq([:html])
+    end
+  end
+
+  it "Loading a nested slice with its own defined slice class and a parent slice's defined slice class prefers the app's defined slice class" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/app.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "config/slices/nested.rb", <<~RUBY
+        require "hanami"
+
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :csv
+            end
+          end
+        end
+      RUBY
```

Comment:
This one actually surprised me! I'm not sure we want this. Because `nested` is not this slice's actual “fully qualified” name, IMO; I'd expected it to be `main/nested`.

Nesting of slices is already complex enough. If possible, I think for now we should eliminate this particular method of falling back to the "grandparent" (or "great parent" or so on for even deeper nesting) for loading a slice's class file.

So, given a main->nested slice class, here's where I'd expect us to allow its slice class to be defined:

1. `slices/main/config/slices/nested.rb`
2. `slices/main/slices/nested/config/slice.rb`

There is no step three. Same applies for any arbitrary level of nesting: load the slice class by it's parent's `config/slices/[slice_name].rb`, or by its own internal `config/slice.rb` only.

What do you think? We could always extend this later, but given how niche and potentially complex nested slices are, I'd prefer to keep the rules simpler for starters.

I think this might have the beneficial side effect of allowing you to simplify or even remove the need for the `SliceRegistrar#ancestors` method?

What do you think?

Reply:
Yeah, I'm totally fine with adding this limitation, I would want this order of precedence
- <app>/config/slices/main/nested.rb
- slices/main/config/slices/nested.rb
- slices/main/slices/nested/config/slice.rb

so the app's file loads ahead of the enclosing slice's file which loads ahead of the actual slice's file.